### PR TITLE
feat: fill form by query params

### PR DESCRIPTION
**Pull request by rallisf1**
ref: #1371
closes: #2862
~~Work In Progress~~

## Change Summary

Added Prefill mode select in Form Share View with 4 options:

- Allow pre-filling fields: parse fields from query params and fields remain editable
- Disable pre-filling fields: disable parsing fields from query params
- Lock pre-filled fields as read-only: parse fields from query params and fields become read-only
- Hide pre-filled fields: parse fields from query params and fields become invisible

__Note: My use-case didn't require per field options so the options above work for all the fields in the form__

WIP:

- Also works for "Has Many / Many To Many / Belongs To", ~~probably more virtual fields~~ but it doesn't populate the value list. I'd appreciate someone to step in on this
- ~~i18n, will do once this is finalized and accepted~~

### How to use

Add `&fieldname=value` in the query params of the form url. Spaces and special characters work as expected in modern browsers (they handle the URL parsing). Special fields:

- checkbox=1
- multiselect=option1,option2...
- date=2025-01-10 `only accepts ISO format`
- duration=18000 `must be in milliseconds`
- hasmany=id1|Title1;id2|Title2 `| splits id from title and ; splits multiple values`

## Change type

- [X] feat: (new feature for the user, not a new feature for build script)
- [ ] fix: (bug fix for the user, not a fix to a build script)
- [ ] docs: (changes to the documentation)
- [ ] style: (formatting, missing semi colons, etc; no production code change)
- [ ] refactor: (refactoring production code, eg. renaming a variable)
- [ ] test: (adding missing tests, refactoring tests; no production code change)
- [ ] chore: (updating grunt tasks etc; no production code change)

## Test/ Verification

Provide summary of changes.

## Additional information / screenshots (optional)

d450b5dd-1472-4a46-b2fa-48c900aff708

### CODE REVIEW

**Review by pranavxc — APPROVED** [review]

**Review by wingkwong — CHANGES_REQUESTED** [review]
1. Prefer the following layout. Since users cannot construct the query param on UI, next to Pre-filled Fields, we may need an indicator / tooltip to show the details (e.g. what are those 4 options / how to construct the query params / what column types are currently supported etc). Ideally it should be a link to the documentation. 
image
2. So far it looks good. Please proceed to add i18n to `en.json`
3. Not working in survey mode
4. Run `npm run lint` on `nc-gui` after all the changes.

File: packages/nc-gui/pages/[projectType]/form/[viewId]/index/index.vue [hunk 1]
```diff
@@ -5,9 +5,32 @@ import { ref } from 'vue'
...
+if (formColumns.value && sharedViewMeta.value.preFilledMode !== 'none' && Object.keys(route.query).length > 0) {
+  for (const column of formColumns.value) {
+    if (column.title && route.query[column.title]) {
+      if (isVirtualCol(column)) {
+        const virtData = (route.query[column.title] as string).split(';')
+        if (virtData.length) {
+          formState.value[column.title] = []
+          virtData.forEach((virtItem) => {
+            const virtRow = virtItem.split('|')
+            formState.value[column.title] = [...formState.value[column.title], { Id: virtRow[0], Title: virtRow[1] }]
```
Comment (pranavxc) [inline packages/nc-gui/pages/[projectType]/form/[viewId]/index/index.vue:22]: The `Id` nd `Title` should be replaced with the primary column and display column title of the related table. You can extract these values from the meta.

Comment (rallisf1) [reply packages/nc-gui/pages/[projectType]/form/[viewId]/index/index.vue:22]: OMG you're right. Not 100% familiar with the project yet; forgive my noob mistakes. Will finalize this PR during the weekend. Peace.

Comment (pranavxc) [reply packages/nc-gui/pages/[projectType]/form/[viewId]/index/index.vue:22]: And for belongs to it should be an object

**Review by dstala — COMMENTED** [review]
- verified except for LinkToAnotherRecord; we can defer pre-filling of links in the first iteration.
- [to be discussed; low priority] constructing URL is currently manual process. we can evaluate to add Generate pre-filled URL button. One can configure statically pre-filled value fields in form view & generate link. This button can be part of shared view modal.

**Comment by rallisf1** [comment]
This is finalized from my end.

To sum up:

- This doesn't work in Survey mode and it acts like that in the UI and is mentioned in the Docs.
- [Bug] The Virtual Cells save the pre-filled value (if provided) but don't populate their value list. I have no clue how to fix this.
- i18n and all changes required have been made.

I want to thank @wingkwong for all his help and of course the rest of you for your inputs.

**Comment by wingkwong** [comment]
Please resolve the conflicting files.

**Comment by wingkwong** [comment]
Will review again after the next release.

**Review by dstala — COMMENTED** [review]
Looks like form field for select fields (single select and multi select) is broken after rebase. I tried it out with following configuration, meta db project. Was unable to input select fields from shared form.

Node: **v16.14.0**
Arch: **arm64**
Platform: **darwin**
Docker: **false**
RootDB: **sqlite3**
PackageVersion: **0.109.1**

I will continue to test this feature further & will capture additional here as and when I find.

**Comment by dstala** [comment]
@rallisf1 Can you help rebase this PR to latest develop?

**Comment by rallisf1** [comment]
> @rallisf1 Can you help rebase this PR to latest develop?

Kinda swamped this week but I'll try.

**Comment by dstala** [comment]
@rallisf1 thanks for this PR. as this was old, we prepared a separate PR based out of your proposal here & merged into develop.

**Comment by rallisf1** [comment]
Gee thanks! I never got to updating this as you were refactoring a bit too fast for me to catch up.

**Pull request closed without merging**